taskgraph/Task.py

```diff
@@ -898,20 +822,35 @@ def _terminate(self):
             "Invoking terminate. already terminated? %s", self._terminated)
         if self._terminated:
             return
-        self._terminated = True
-
-        if self._logging_queue:
-            self._logging_queue.put(None)
+        try:
+            # it's possible the global state is not well defined, so just in
+            # case we'll wrap it all up in a try/except
+            self._terminated = True
+            if self._executor_ready_event is not None:
+                # alert executors to check that _terminated is True
+                self._executor_ready_event.set()
+            LOGGER.debug("shutting down workers")
+            if self._worker_pool is not None:
+                self._worker_pool.close()
+                self._worker_pool.terminate()
+                self._worker_pool = None
 
-        for task in self._task_hash_map.values():
-            LOGGER.debug("setting task done for %s", task.task_name)
-            task.task_done_executing_event.set()
+            # This will terminate the logging worker
+            if self._logging_queue is not None:
+                self._logging_queue.put(None)
 
-        if self._worker_pool:
-            self._worker_pool.close()
-            self._worker_pool.terminate()
+            # This will cause all 'join'ed Tasks to join.
+            if self._n_workers >= 0:
+                self._executor_ready_event.set()
+                if self._reporting_interval is not None:
+                    self._execution_monitor_wait_event.set()
+                for task in self._task_hash_map.values():
+                    # shortcut to get the tasks to mark as joined
+                    task.task_done_executing_event.set()
 
-        self._executor_ready_event.set()
+            LOGGER.debug('taskgraph terminated')
+        except Exception:
+            LOGGER.exception('exception occurred during __del__')

```

Comment:
`"exception occurred during _terminate()"` now that this block and exception handler have been moved out of `__del__`?

Reply:
Fixed in ab146f6